pytorch_lightning/strategies/fully_sharded_native.py

```diff
@@ -0,0 +1,254 @@
+        self.setup_precision_plugin()
+
+    def model_to_device(self) -> None:
+        # ensure we update the device type in the lightning module
+        log.info(f"{self.__class__.__name__}: moving model to device [{self.root_device}]...")
+        self.lightning_module.to(self.root_device)
+
+    @contextlib.contextmanager
+    def model_sharded_context(self) -> Generator:
+        log.detail(f"{self.__class__.__name__}: entered model_sharded_context.")
+
+        with enable_wrap(
+            wrapper_cls=FullyShardedDataParallel,
+            process_group=self.process_group,
+            cpu_offload=self.cpu_offload,
+            backward_prefetch=self.backward_prefetch,
+        ):
+            yield
+
+    def barrier(self, name: Optional[str] = None) -> None:
+        if not distributed_available():
+            return
+        if torch.distributed.get_backend() == "nccl":
+            torch.distributed.barrier(device_ids=self._determine_device_ids())
+        else:
+            torch.distributed.barrier()
+
+    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
+        obj = [obj]
+        if self.global_rank != src:
+            obj = [None]
+        torch.distributed.broadcast_object_list(obj, src, group=_group.WORLD)
+        return obj[0]
+
+    def reduce(
+        self,
+        tensor: Union[torch.Tensor, Any],
+        group: Optional[Any] = None,
+        reduce_op: Optional[Union[ReduceOp, str]] = "mean",
+    ) -> torch.Tensor:
+        """Reduces a tensor from several distributed processes to one aggregated tensor.
+
+        Args:
+            tensor: the tensor to sync and reduce
+            group: the process group to gather results from. Defaults to all processes (world)
+            reduce_op: the reduction operation. Defaults to 'mean'/'avg'.
+                Can also be a string 'sum' to calculate the sum during reduction.
+
+        Return:
+            reduced value, except when the input was not a tensor the output remains is unchanged
+        """
+        if isinstance(tensor, torch.Tensor):
+            tensor = sync_ddp_if_available(tensor, group, reduce_op=reduce_op)
+        return tensor
+
+    def _determine_device_ids(self) -> Optional[List[int]]:
+        return [self.root_device.index]
+
+    def teardown(self) -> None:
+        log.info(f"{self.__class__.__name__}: tearing down strategy...")
+        super().teardown()
+
+        if (
+            self.lightning_module.trainer is not None
+            and self.lightning_module.trainer.state.fn == TrainerFn.FITTING
+            and self._layer_sync
+        ):
+            self.model = self._layer_sync.revert(self.model)
+
+        if self.root_device.type == "cuda":
+            # GPU teardown
+            log.info(f"{self.__class__.__name__}: moving model to CPU...")
+            self.lightning_module.cpu()
+            # clean up memory
+            torch.cuda.empty_cache()
```

Comment:
should we be doing this, even for CPU offload? I remember when we first did the FSDP integration with fairscale, this lead to OOMs because of the .cpu call

Reply:
Currently, this also happens for the FairScale FSDP Strategy too. it makes sense that this is a bad idea especially if the model was created in the `configure_sharded_model` context (assuming creation on GPU).

I think what makes the most sense is leaving the model on whatever device was selected during training. I assume in most power user cases you'll be just training within the script/function, not trying to do anything with the model other than potentially save.

@ananthsub would you be up for removing this logic from the teardown? We should consider doing the same for fully sharded fairscale as well.